### DISCUSSION TITLE
Bitmap vs WriteableBitmap confusion eg. saving JPEG

### DISCUSSION THREAD

**Discussion by buzzware** [AvaloniaUI/Avalonia#15969] [General]
My app needs to load and save images, and display thumbnails. So I began using Bitmap which loads ".jpg" fine.
But when I needed to save JPEG thumbnails it got difficult, and the solutions are wasteful in processor and memory.
Strangely, Bitmap.Save only saves in PNG format, even though Bitmap.Load loads both PNG & JPEG.
Underneath, Skia supports JPEG saving like this `skImage.Encode(SKEncodedImageFormat.Jpeg, 90)`.
But how do you get a SKImage from a Bitmap? It seems you have to do Bitmap->WriteableBitmap-> ILockedFramebuffer->SKBitmap->SKImage (see below).
To avoid one step I could use WriteableBitmap everywhere, but that will cause problems one day, isn't intuitive and shouldn't be necessary.
1) Shouldn't Bitmap be the default bitmap class for most code that simply loads and saves bitmaps ?
2) Shouldn't Bitmap support Save("pic.jpg",90) or similar ? I actually want to save to a memory stream, and so I don't actually like specifying a filename in order to specify the format.
3) WriteableBitmap.Lock() returns a ILockedFramebuffer - perhaps Bitmap could return a new read-only ILockedFramebuffer without the need for locking, and then it could be used for direct read-only pixel operations.
4) Is there a better workaround than below ?

```
// !!! This was AI generated - I haven't even tested it, but it does compile

var writeableBitmap = new WriteableBitmap(avaloniaBitmap.PixelSize);

// Draw Avalonia bitmap to writablebitmap
using (var context = new DrawingContext(writeableBitmap.Item))
{
    context.DrawImage(avaloniaBitmap, new Rect(avaloniaBitmap.Size), new Rect(avaloniaBitmap.Size));
}

// Lock FB and convert to SKBitmap
using (var fbContext = writeableBitmap.Lock())
{
    var skiaImageInfo = new SKImageInfo(writeableBitmap.PixelSize.Width, writeableBitmap.PixelSize.Height);
    var skiaBitmap = new SKBitmap(skiaImageInfo);

    unsafe
    {
        var skiaPixels = (SKColor*)fbContext.Address.ToPointer();
        skiaBitmap.SetPixels((IntPtr)skiaPixels);
    }
    
    // create SKImage from SKBitmap
    var skImage = SKImage.FromBitmap(skiaBitmap);
    using (var stream = File.OpenWrite("/path/to/your/output.jpg")) {
	skImage.Encode(SKEncodedImageFormat.Jpeg, 90).SaveTo(stream);
    }
}
```

**Comment by thevortexcloud** [comment]
> But how do you get a SKImage from a Bitmap? 

You can write a custom bitmap implementation that exposes all the Skia stuff.

https://github.com/AvaloniaUI/Avalonia/discussions/13610#discussioncomment-7576208

Avalonia's bitmap implementation is an abstraction around what ever backend you are using. Generally you would use the Skia backend, but there are other backends supported on some platforms so it does not make sense for it to directly expose the Skia types.

**Comment by stevemonaco** [comment]
Agree with skipping `Bitmap` altogether since you're directly coupled anyways. Just use `SKBitmap` and render it directly. It's fairly trivial and you can look at https://github.com/wieslawsoltes/Svg.Skia/blob/master/src/Avalonia.Controls.Skia/SKBitmapControl.cs for reference.

Generally, the UI framework shouldn't own your state and `Bitmap` is another example of this (IMO). It should only be used for display and it's a bit weird that `Save` exists on it at all.

**Comment by buzzware** [comment]
For app development I'm happy with a default readonly Bitmap class, but I think there should be :

1) a way to load and save at least PNG & JPEG, even with an external class or static method. 
2) both sync (low level, run it on an alternate thread) & async (for convenience, run it on any thread and it won't block the UI decoding big files) methods for streams and files
3) read only direct pixel access methods

Then a WriteableBitmap class extending Bitmap makes sense to me.

With the above I would only need for Skia for those 5% cases requiring fancy drawing.

**Reply by buzzware** [reply]
It seems achievable to write and use custom bitmap classes. I don't see any compatibility issues with the rest of the framework. You can't implement IBitmap because it is private, which also means there is no need to. It seems the only interfaces required are IImageBrushSource, IImage & IDisposable.

**Reply by stevemonaco** [reply]
I imagine there was some oversight or difficulty in putting JPEG support. Remember that if it's supported, then every backend requires it (not just the Skia backend). I'm not sure there are any production quality managed libraries for this besides ImageSharp, else you'd have to ship native dependencies for ImageMagick or some libjpeg variant...which adds an unacceptable amount of extra binary size.

There's not really a point in using `IBitmap`. Just implement `IImage` and `IDisposable` (you'll need to manage disposing yourself), use `<Image />`, and create+assign the `Source` in code-behind.

**Comment by buzzware** [comment]
I'm having success with these helper methods :

```
	public static byte[] EncodeThumbnailJpeg(Bitmap bitmap, int width, int height, int quality) {
		var resized = ThumbnailBitmapFrom(bitmap, width, height);
		return EncodeToJpeg(resized, quality);
	}

	public static Bitmap DecodeBitmap(IEnumerable<byte> blob) {
		using var stream = new MemoryStream(blob.ToArray());
		return new Bitmap(stream);
	}
	
	public static Bitmap ThumbnailBitmapFrom(Bitmap bitmap,int width, int height) {
		var scale = Math.Min(width / (double)bitmap.Size.Width, height / (double)bitmap.Size.Height);
		var scaledWidth = (int)(bitmap.Size.Width * scale);
		var scaledHeight = (int)(bitmap.Size.Height * scale);
		var resized = bitmap.CreateScaledBitmap(new PixelSize(scaledWidth, scaledHeight), BitmapInterpolationMode.MediumQuality);
		return resized;
	}

	public static WriteableBitmap WriteableFromBitmap(Bitmap bitmap) {
		var writeableBitmap = new WriteableBitmap(
			bitmap.PixelSize,
			bitmap.Dpi,
			bitmap.Format
		);
		using (var fb = writeableBitmap.Lock()) {
			bitmap.CopyPixels(fb,AlphaFormat.Opaque);
		}
		return writeableBitmap;
	}

	private static Dictionary<PixelFormat, SKColorType> ColorTypeMap = new Dictionary<PixelFormat, SKColorType>() {
		[PixelFormat.Bgra8888] = SKColorType.Bgra8888
	};
	
	public static byte[] EncodeToJpeg(Bitmap bitmap, int quality = 80) {
		WriteableBitmap? wb = bitmap as WriteableBitmap;
		WriteableBitmap? tempWb = (wb == null) ? WriteableFromBitmap(bitmap) : null;
		wb ??= tempWb;
		try {
			using var pi = wb!.Lock();
			SKColorType colorType = ColorTypeMap[bitmap.Format!.Value];
			var skImageInfo = new SKImageInfo(pi.Size.Width, pi.Size.Height, colorType);
			using var skBitmap = new SKBitmap(skImageInfo);
			skBitmap.InstallPixels(skImageInfo, pi.Address, pi.RowBytes);
			using var skImage = SKImage.FromBitmap(skBitmap);
			return skImage.Encode(SKEncodedImageFormat.Jpeg, quality).ToArray();
		} finally {
			tempWb?.Dispose();
		}
	}
```
